Sources/SwiftFormatPrettyPrint/OperatorContext.swift

```diff
@@ -0,0 +1,368 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Contains information about precedence groups and operators used when folding
+/// sequence expressions.
+public final class OperatorContext {
+
+  /// The key used to store the associativity between two precedence groups in
+  /// the cache.
+  private struct AssociativityCacheKey: Equatable, Hashable {
+    private let first: ObjectIdentifier
+    private let second: ObjectIdentifier
+
+    init(_ first: PrecedenceGroup, _ second: PrecedenceGroup) {
+      self.first = ObjectIdentifier(first)
+      self.second = ObjectIdentifier(second)
+    }
+  }
+
+  /// A dictionary of known precedence groups, keyed by their names.
+  private var precedenceGroups = [PrecedenceGroup.Name: PrecedenceGroup]()
+
+  /// A dictionary of known infix operators, keyed by their name.
+  private var infixOperators = [String: InfixOperator]()
+
+  /// A cache of associativities between pairs of operators, computed and cached
+  /// on demand.
+  private var associativityCache = [AssociativityCacheKey: Associativity?]()
+
+  /// Creats a new operator context that is populated with the precedence groups
+  /// operators that are built into the Swift compiler and standard library.
+  public init() {
+    // Add the built-in precedence groups.
```

Comment:
Nitpick, I'd recommend to decouple a bit `OperatorContext` from how to populate it. So for example there could be an `addDefaultOperators` function and for convenience a `getDefaultOperatorContext` static function. Later on we could potentially have a function to populate an `OperatorContext` by loading a module (e.g. the stdlib) and for this case not need the hardcoded operators.

Reply:
`Builtin` as in `addBuiltinOperators` is slightly better.

Reply:
Done.